libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthSessionRefresher.m

```diff
@@ -95,13 +96,35 @@ - (void)completeWithSuccess:(SFOAuthCredentials *)credentials {
 
 - (void)completeWithError:(NSError *)error {
     [self log:SFLogLevelError format:@"%@ Refresh failed with error: %@", NSStringFromSelector(_cmd), error];
-    if (self.errorBlock) {
+
+    if ([error.domain isEqualToString:kSFOAuthErrorDomain] && error.code == kSFOAuthErrorInvalidGrant) {
+        [self log:SFLogLevelInfo format:@"%@ Invalid grant error received, triggering logout.", NSStringFromSelector(_cmd)];
+        // make sure we call logoutUser on main thread
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [self createAndStoreLogoutEvent:error];
+            [self logoutUser];
+        });
+    }
+    else if (self.errorBlock) {
         dispatch_async(dispatch_get_main_queue(), ^{
             self.errorBlock(error);
         });
     }
 }
 
+- (void)logoutUser {
+    SFUserAccountIdentity *userIdentity = [[SFUserAccountIdentity alloc] initWithUserId:self.coordinator.credentials.userId orgId:self.coordinator.credentials.organizationId];
```

Comment:
This is kind of going in the wrong direction from this class's original scope.  The idea is that this class is merely a conduit for the OAuth negotiations, and its consumer would ultimately be responsible for the results.

I think this functionality should be handled upstream of this class.